src/Microsoft.ML.Transforms/Text/NgramTransform.cs

```diff
@@ -843,7 +843,7 @@ internal static bool IsColumnTypeValid(ColumnType type)
             if (!(type.ItemType is KeyType itemKeyType))
                 return false;
             // Can only accept key types that can be converted to U4.
-            if (itemKeyType.Count == 0 && type.ItemType.RawKind > DataKind.U4)
+            if (itemKeyType.Count == 0 && !NgramUtils.IsValidNgramRawType(itemKeyType.RawType))
```

Comment:
>IsValidNgramRawType(itemKeyType.RawType [](start = 54, length = 39)

If I read the comment, it says, `Can only accept key types that can be converted to U4,` in every place that this `IsValidNgramRawType` is being called. That basically means U1, U2, and U4 are acceptable, and only U8 is not. So I'd have expected, *based on this comment*, that this would be a test against whether the raw type is a `ulong` or not.

*However*, if I then go and read this `IsValidNgramRawType`, it is doing something else! It is checking against all sorts of types that have nothing to do with key-types *at all*, including `sbyte`, `short`, `int`... which makes no sense to me, given the comment, the fact that these are *definitely* meant to operate over keytypes, etc.

I would have expected just making sure it isn't a `ulong` would have been sufficient, but...

Reply:
I was just converting the existing logic: `type.ItemType.RawKind > DataKind.U4`.

I can change that method to assume that `KeyType`s can only be `byte`, `ushort`, `uint`, or `ulong`, and only make it check for `ulong`.

---
In reply to: [248037058](https://github.com/dotnet/machinelearning/pull/2143#discussion_r248037058) [](ancestors = 248037058)